### PR TITLE
JIT: Remove some `fgRenumberBlocks` calls

### PR DESCRIPTION
**Pull request by amanasifkhalid**
Part of #107749. While we cannot get rid of `fgRenumberBlocks` altogether yet, this gets us close. The JIT still as a few hard dependencies on bbNums reflecting relative lexical order:
* If the CFG has loops, `optSetBlockWeights` will use bbNums for naive loop backedge detection.
* Loop inversion also uses bbNums in lieu of flow edges to guess at a loop's shape.
* Loop duplication helpers iterate the loop's blocks in lexical order, which requires ordered bbNums. As such, I've left the `fgRenumberBlocks` calls in loop cloning/unrolling in-place for now.

There's also a soft dependency on bbNum ordering in `fgUpdateFlowGraph` that can trigger diffs. These diffs were quite small locally, so I think we can tolerate them for now.

### CODE REVIEW

File: src/coreclr/jit/compiler.cpp [hunk 1]
```diff
@@ -4928,6 +4921,15 @@ void Compiler::compCompile(void** methodCodePtr, uint32_t* methodCodeSize, JitFl
 
             if (doLoopHoisting)
             {
+                if (doBranchOpt)
+                {
+                    // bbNum quirk: RBO uses the current BB epoch size to determine if loop hoisting created blocks.
+                    // Ensure the BB epoch is up to date before running loop hoisting to avoid false positives.
+                    // TODO-bbNum: Either replace this check with a breadcrumb like 'fgModified',
+                    // or relax jump threading's prerequisites.
+                    EnsureBasicBlockEpoch();
```
Comment (amanasifkhalid) [inline src/coreclr/jit/compiler.cpp:4930]: On second thought, we should be able to get rid of RBO's dependency on block epoch altogether. Let me do that first.

**Comment by amanasifkhalid** [comment]
cc @dotnet/jit-contrib, @jakobbotsch PTAL. [Small diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=875933&view=ms.vss-build-web.run-extensions-tab). The Instrumented Tier0 diffs stem from the fact that bbNums are used in key computations by `EfficientEdgeCountBlockToKey` [here](https://github.com/dotnet/runtime/blob/c31bb93ea3ca685b6336cd0cbde4893dff30f80c/src/coreclr/jit/fgprofile.cpp#L1174), so the lack of renumbering is churning the instrumentation schema a bit. The resultant codegen diffs are from the address of the profile data now fitting in 32/64 bits, instead of 64/32 bits. Thanks!

**Comment by AndyAyersMS** [comment]
> The Instrumented Tier0 diffs stem from the fact that bbNums are used in key computations by `EfficientEdgeCountBlockToKey`

We can change this to use `bbID` too. It may break reading in static PGO data for a while but so does changing the values of `bbNum`. Note when we get around to https://github.com/dotnet/runtime/issues/44372, instrumented inlinees will need to adjust this value to make it inlinee-relative (as if we'd instrumented the method as root).

**Comment by amanasifkhalid** [comment]
> We can change this to use bbID too.

Do we have anything to gain from switching this over to `bbID`? Assuming `bbNum` continues to stick around, it seems like it would be easier to keep using it for the block key since it's already inlinee-relative, right?

**Comment by AndyAyersMS** [comment]
> > We can change this to use bbID too.
> 
> Do we have anything to gain from switching this over to `bbID`? Assuming `bbNum` continues to stick around, it seems like it would be easier to keep using it for the block key since it's already inlinee-relative, right?

`bbID` seems more stable, but I suppose as long as we're not ever tempted to add renumbering early, this can stay as is.

**Review by jakobbotsch — APPROVED** [review]

**Pull request merged**